### PR TITLE
Pair.toPairs accepts arrays

### PR DESCRIPTION
**Pull request by aecepoglu**
Pair.toPairs() can actually work on arrays without needing any modifications.

This PR simply toggles on the ability and adds tests for it.

It is dependent or an already open PR: https://github.com/evilsoft/crocks/pull/356

### CODE REVIEW

**Comment by dalefrancis88** [comment]
Could this be done to not need that other PR? it looks like it

**Comment by aecepoglu** [comment]
It could but the other PR is already within the list of issues and I assume will get merged.

File: src/Maybe/find.spec.js [hunk 1]
```diff
@@ -67,13 +64,11 @@ test('find works with predicates', t => {
 
   t.equal(find(largeNumber, [ 10, '12', 150, 200, 2000 ]).option(0), 150, 'works with predicates')
   t.ok(isSameType(Maybe, isTrue([])), 'returns a Maybe for array value')
-  t.ok(isSameType(Maybe, isTrue(fromArray([]))), 'returns a Maybe for List value')
```
Comment (dalefrancis88) [inline src/Maybe/find.spec.js]: we still need this test, this was being used to test that `find` worked with `List`

File: src/Maybe/find.spec.js [hunk 2]
```diff
@@ -67,13 +64,11 @@ test('find works with predicates', t => {
...
   t.equal(gt3([ 1, 2, 3, 4, 5, 6 ]).option('Nothing'), 4, 'returns the correct value')
   t.equal(gt3([ 0, null, undefined, 4, 5, 6 ]).option('Nothing'), 4, 'handles bad values in foldable')
   t.equal(gt3([ 1, 2, 3 ]).option('Nothing'), 'Nothing', 'returns nothing when value is not found')
   t.equal(isTrue([]).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items ([])')
   t.equal(isTrue(new Array()).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (Array)')
-  t.equal(isTrue(fromArray([])).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (List)')
```
Comment (dalefrancis88) [inline src/Maybe/find.spec.js]: Same here

File: src/Maybe/find.spec.js [hunk 3]
```diff
@@ -88,15 +83,13 @@ test('find works as expected', t => {
   t.ok(isFunction(find), 'is a function')
 
   t.ok(isSameType(Maybe, isTrue([])), 'returns a Maybe for array value')
-  t.ok(isSameType(Maybe, isTrue(fromArray([]))), 'returns a Maybe for List value')
```
Comment (dalefrancis88) [inline src/Maybe/find.spec.js]: Same here

File: src/Maybe/find.spec.js [hunk 4]
```diff
@@ -88,15 +83,13 @@ test('find works as expected', t => {
...
 
   t.equal(gt3([ 1, 2, 3, 4, 5, 6 ]).option('Nothing'), 4, 'returns the correct value')
 
   t.equal(gt3([ 0, null, undefined, 4, 5, 6 ]).option('Nothing'), 4, 'handles bad values in foldable')
   t.equal(gt3([ 1, 2, 3 ]).option('Nothing'), 'Nothing', 'returns nothing when value is not found')
   t.equal(isTrue([]).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items ([])')
   t.equal(isTrue(new Array()).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (Array)')
-  t.equal(isTrue(fromArray([])).option('Nothing'), 'Nothing', 'returns a Nothing when there are no items (List)')
```
Comment (dalefrancis88) [inline src/Maybe/find.spec.js]: Same Here

File: src/Pair/toPairs.js [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
-const List = require('../core/List')
 const Pair = require('../core/Pair')
 const isObject = require('../core/isObject')
+const isArray = require('../core/isArray')
 
 // toPairs : Object -> List (Pair String a)
 function toPairs(obj) {
-  if(!isObject(obj)) {
+  if(!isObject(obj) && !isArray(obj)) {
     throw new TypeError('toPairs: Object required for argument')
```
Comment (dalefrancis88) [inline src/Pair/toPairs.js]: 'toPairs: Object or Array required for argument'

File: src/Pair/toPairs.spec.js [hunk 6]
```diff
@@ -57,3 +55,41 @@ test('toPairs', t => {
...
+    null,
+    0,
+    1,
+    '',
+    'string',
+    false,
+    true,
+    {},
+    []
+  ].map((x, i) => ({
```
Comment (dalefrancis88) [inline src/Pair/toPairs.spec.js]: I'm not sure i am the biggest fan of this formatting. Could we clean it up to be a bit more pointfree, and lets not do it straight from the array

**Comment by aecepoglu** [comment]
should be OK now.

**Comment by dalefrancis88** [comment]
Hey @evilsoft are you happy for this to be merged in preparation for the 0.12.0 release?

File: docs/src/pages/docs/crocks/Pair.md [hunk 7]
```diff
@@ -887,14 +887,12 @@ const record = {
 const joinField = (key, value) =>
   `${key}:${value}`
 
-// joinRecord :: List String -> String
-const joinRecord = list =>
-  list.toArray()
-    .join('|')
+// joinRecord :: String -> [String] -> String
+const join = separator => array => array.join(separator)
```
Comment (dalefrancis88) [inline docs/src/pages/docs/crocks/Pair.md]: `array.join(separator)` should be on a second line

Comment (dalefrancis88) [reply docs/src/pages/docs/crocks/Pair.md]: I believe we could also simplify this to `const join = array => array.join`

File: src/Pair/toPairs.js [hunk 8]
```diff
@@ -1,21 +1,21 @@
...
 
 // toPairs : Object -> List (Pair String a)
 function toPairs(obj) {
-  if(!isObject(obj)) {
-    throw new TypeError('toPairs: Object required for argument')
+  if(!isObject(obj) && !isArray(obj)) {
+    throw new TypeError('toPairs: Object or Array required for argument')
   }
 
   return Object.keys(obj).reduce(
```
Comment (dalefrancis88) [inline src/Pair/toPairs.js]: can we re-write this using `Object.entries`

```javascript
(acc, [ key, value ]) => ...
```

File: docs/src/pages/docs/crocks/Pair.md (hunk 7 above)
Comment (aecepoglu) [reply docs/src/pages/docs/crocks/Pair.md]: done.

File: src/Pair/toPairs.js (hunk 8 above)
Comment (aecepoglu) [reply src/Pair/toPairs.js]: done.

I did the changes without access to a coding environment, so apologies for the high number of commits.

**Pull request closed without merging**